context-monitoring/src/main/java/de/atb/context/monitoring/monitors/file/AbstractFileSystemMonitor.java

```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    protected final void stopWatcher() {
+        if (this.watchDaemon != null) {
+            this.watchDaemon.interrupt();
+            try {
+                this.watchDaemon.join(2000);
+            } catch (InterruptedException e) {
+                this.logger.warn(e.getMessage(), e);
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    protected final void prepareWatcher() {
+        final File finalPath = this.pathToMonitor;
+        Thread t = new Thread(() -> iterateFiles(finalPath), this.getClass().getSimpleName() + " Iterator Thread");
+        t.setDaemon(true);
+        t.start();
+    }
+
+    protected final void startWatcher() {
+        prepareWatcher();
+        final File pathToMonitor = this.pathToMonitor;
+        this.watchDaemon = new Thread(new Runnable() {
+            private final Logger logger = LoggerFactory.getLogger(getClass());
+
+            @Override
+            public void run() {
+                WatchService watchService = null;
+                try {
+                    watchService = FileSystems.getDefault().newWatchService();
+                } catch (IOException e) {
+                    logger.error(e.getMessage(), e);
+                }
+                Path watchedPath = Paths.get(pathToMonitor.getAbsolutePath());
+                this.logger.debug("Started monitoring '" + watchedPath + "'");
+                WatchKey signalledKey;
+                try {
+                    watchedPath.register(watchService,
+                        StandardWatchEventKinds.ENTRY_CREATE,
+                        StandardWatchEventKinds.ENTRY_MODIFY,
+                        StandardWatchEventKinds.ENTRY_DELETE,
+                        StandardWatchEventKinds.OVERFLOW);
+                    while (true) {
+                        try {
+                            signalledKey = watchService.take();
+                            Long time = System.currentTimeMillis();
+                            handleWatchEvents(signalledKey.pollEvents(), time);
+                            signalledKey.reset();
+                        } catch (InterruptedException ix) {
+                            this.logger
+                                .info("Watch service was interrupted, closing...");
+                            this.logger.debug(ix.getMessage(), ix);
+                            watchService.close();
+                            Thread.currentThread().interrupt();
+                            break;
+                        } catch (ClosedWatchServiceException cwse) {
+                            this.logger
+                                .info("Watch service closed, terminating...");
+                            this.logger.debug(cwse.getMessage(), cwse);
+                            break;
+                        }
+                    }
+                } catch (IOException e) {
+                    this.logger.error(e.getMessage(), e);
```

Comment:
Shouldn't there be some kind of exception handling besides logging?

Like stopping the monitor (see `shutdown` method)? Or maybe something similar to what's done in lines 143 - 155?

Reply:
in case of exception now `shutdown` is called